test/functional/session_controller_test.rb

```diff
@@ -0,0 +1,116 @@
+require "test_helper"
+require "gds_api/test_helpers/account_api"
+
+class SessionsControllerTest < ActionController::TestCase
+  include GdsApi::TestHelpers::AccountApi
+  include SessionHelpers
+
+  context "GET sign-in" do
+    setup do
+      stub_account_api_get_sign_in_url(redirect_path: "/bank-holiday")
+    end
+
+    should "redirect the user to the GOV.UK Account service domain" do
+      get :create, params: { redirect_path: "/bank-holiday", state_id: "state123" }
+      assert_response :redirect
+      assert_equal @response.headers["Location"], "http://auth/provider"
+    end
+
+    should "preserve the _ga tracking parameter if provided" do
+      get :create, params: { redirect_path: "/bank-holiday", state_id: "state123", _ga: "ga123" }
+      assert_equal @response.headers["Location"], "http://auth/provider?_ga=ga123"
+    end
+
+    context "when already logged in" do
+      setup { mock_logged_in_session }
+
+      should "redirect the user to the account manager if the user is already logged in" do
+        get :create, params: { redirect_path: "/bank-holiday", state_id: "state123" }
+        assert_redirected_to Plek.find("account-manager")
+      end
+    end
+  end
+
+  context "GET sign-in/callback" do
+    setup do
+      @root_path = Plek.new.website_root
+    end
+
+    should "redirect to the root path if the :code param is not present" do
+      stub_account_api_validates_auth_response
```

Comment:
`stub_account_api_validates_auth_response` isn't needed in this test, because the redirect happens before the `GdsApi.account_api.validate_auth_response`

Reply:
ah great spot thanks

Reply:
done here: https://github.com/alphagov/frontend/pull/2668/commits/4c01553d45069b6b07b1130b5b2ffd7a5a481540